rudder-core/src/main/scala/com/normation/rudder/repository/jdbc/ExpectedReportsJdbcRepository.scala

```diff
@@ -225,16 +232,148 @@ class FindExpectedReportsJdbcRepository(
+//        reportsAndConfigIds <- tryo ( jdbcTemplate.query(query_subscript_configids_3, ReportAndConfigIdMapper2).asScala.flatten )
+        t1                  =  System.currentTimeMillis
+        _                   =  TimingDebugLogger.debug(s"GetExpectedReports: configIds: ${t1-t0}ms")
+        reportIds           =  Set[Int]() ++ reportsAndConfigIds.map( _._1)
+        t2                  =  System.currentTimeMillis
+        _                   =  TimingDebugLogger.debug(s"GetExpectedReports: map reportsIds: ${t2-t1}ms")
+        expectecReports     <- getExpectedReports(reportIds)
+        t3                  =  System.currentTimeMillis
+        _                   =  TimingDebugLogger.debug(s"GetExpectedReports: expectedreports: ${t3-t2}ms")
       } yield {
-        toExpectedReports(entries).toSet
+
+        val reports = expectecReports.map(x => (x.pkId, x)).toMap
+
+        val mapped = reportsAndConfigIds.map { case (pkId, nodeId, versions) =>
+          val r = reports(pkId)
+          ReportAndNodeMapping(
+             r.pkId
+           , r.nodeJoinKey
+           , r.ruleId
+           , r.serial
+           , r.directiveId
+           , r.component
+           , r.cardinality
+           , r.componentsValues
+           , r.unexpandedCptsValues
+           , r.beginDate
+           , r.endDate
+           , nodeId
+           , versions
+          )}
+
+        val t4 =  System.currentTimeMillis
+        TimingDebugLogger.debug(s"GetExpectedReports: reportsAndNodeMapping: ${t4-t3}ms")
+
+        val res= toNodeExpectedReports(nodeConfigIds, mapped)
+
+        val t5 =  System.currentTimeMillis
+        TimingDebugLogger.debug(s"GetExpectedReports: toNodeExpectedReports: ${t5-t4}ms")
+
+        res
+      }
+
+
+      res
+    }
+  }
+
+  object ReportAndConfigIdMapper extends RowMapper[(Int, NodeId, List[NodeConfigId])] {
+    def mapRow(rs : ResultSet, rowNum: Int) : (Int, NodeId, List[NodeConfigId]) = {
+      def notNullS[T](name: String): String = {
+        rs.getString(name) match {
+          case null => throw new IllegalArgumentException(s"Column '${name}' is null (illegal value)")
+          case x => x
+        }
+      }
+      def notNullT(name: String): Timestamp = {
+        rs.getTimestamp(name) match {
+          case null => throw new IllegalArgumentException(s"Column '${name}' is null (illegal value)")
+          case x => x
+        }
+      }
+
+      (
+          rs.getInt("pkid")
+        , NodeId(notNullS("nodeid"))
+        , NodeConfigVersionsSerializer.unserialize(rs.getArray("nodeconfigids"))
+      )
+    }
+  }
+
+  /**
+   * A specialized version for the query where nodeid is aggredated to the mkString of nodeconfigid.
+   */
+  object ReportAndConfigIdMapper2 extends RowMapper[Seq[(Int, NodeId, List[NodeConfigId])]] {
+    private[this] def unserialize(versions: java.sql.Array): Seq[(NodeId, List[NodeConfigId])] = {
+      if(null == versions) throw new IllegalArgumentException(s"ERROR")
```

Comment:
if there is no previous value (in case of migration), it the array null or is it empty ?